### ISSUE TITLE
matplotlib's plot_contour() method raise an error when given dynamic search space.

### ISSUE THREAD

**Issue by c-bata** [optuna/optuna#3681]
### Expected behavior

As plotly's plot_contour() can handle dynamic search spaces properly, matplotlib's plot_contour() method should be able to accept dynamic search space.

### Environment

- Optuna version:3.0.0rc0.dev
- Python version:3.9.11
- OS:macOS-12.3.1-x86_64-i386-64bit
- (Optional) Other libraries and their versions:

### Error messages, stack traces, or logs

```shell
$ python bug_matplotlib_plot_contour.py
[I 2022-06-16 11:43:25,236] A new study created in memory with name: A single-objective study with 3-dimensional dynamic search space
[I 2022-06-16 11:43:25,238] Trial 0 finished with value: -0.9723216524479252 and parameters: {'category': 'bar', 'x2': -5.986063716221181}. Best is trial 0 with value: -0.9723216524479252.
[I 2022-06-16 11:43:25,239] Trial 1 finished with value: 1.760435351638048 and parameters: {'category': 'foo', 'x1': 3.326813985318985}. Best is trial 1 with value: 1.760435351638048.
[I 2022-06-16 11:43:25,239] Trial 2 finished with value: -7.583657590003812 and parameters: {'category': 'bar', 'x2': -2.246155852266906}. Best is trial 1 with value: 1.760435351638048.
[I 2022-06-16 11:43:25,239] Trial 3 finished with value: 18.474087129547364 and parameters: {'category': 'foo', 'x1': 6.298149267946306}. Best is trial 3 with value: 18.474087129547364.
[I 2022-06-16 11:43:25,240] Trial 4 finished with value: -9.428683789636048 and parameters: {'category': 'bar', 'x2': -1.9293838094551692}. Best is trial 3 with value: 18.474087129547364.
[I 2022-06-16 11:43:25,240] Trial 5 finished with value: -15.02901036791338 and parameters: {'category': 'bar', 'x2': -8.876726759511609}. Best is trial 3 with value: 18.474087129547364.
[I 2022-06-16 11:43:25,240] Trial 6 finished with value: 35.83677106510998 and parameters: {'category': 'foo', 'x1': 7.9863821349050195}. Best is trial 6 with value: 35.83677106510998.
[I 2022-06-16 11:43:25,241] Trial 7 finished with value: 38.300159300048314 and parameters: {'category': 'foo', 'x1': 8.188712248929361}. Best is trial 7 with value: 38.300159300048314.
[I 2022-06-16 11:43:25,241] Trial 8 finished with value: 36.79416883099965 and parameters: {'category': 'foo', 'x1': 8.065819716328507}. Best is trial 7 with value: 38.300159300048314.
[I 2022-06-16 11:43:25,241] Trial 9 finished with value: 41.966519235944105 and parameters: {'category': 'foo', 'x1': 8.478157086389933}. Best is trial 9 with value: 41.966519235944105.
[I 2022-06-16 11:43:25,251] Trial 10 finished with value: 2.9905738109692943 and parameters: {'category': 'foo', 'x1': 0.2706724396548541}. Best is trial 9 with value: 41.966519235944105.
[I 2022-06-16 11:43:25,253] Trial 11 finished with value: 63.81133568419859 and parameters: {'category': 'foo', 'x1': 9.988199777434124}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,256] Trial 12 finished with value: 59.22688123709178 and parameters: {'category': 'foo', 'x1': 9.695900287626639}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,258] Trial 13 finished with value: 61.576558077718104 and parameters: {'category': 'foo', 'x1': 9.847073217303258}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,260] Trial 14 finished with value: 61.105536719061355 and parameters: {'category': 'foo', 'x1': 9.817003052261228}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,263] Trial 15 finished with value: 11.129176866576213 and parameters: {'category': 'foo', 'x1': 5.336042096043785}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,266] Trial 16 finished with value: 20.299871591942505 and parameters: {'category': 'foo', 'x1': 6.505537880424767}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,268] Trial 17 finished with value: 2.3066620688215416 and parameters: {'category': 'foo', 'x1': 3.518769919646008}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,270] Trial 18 finished with value: -23.647267561545284 and parameters: {'category': 'bar', 'x2': -9.862845623865237}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,273] Trial 19 finished with value: 59.04950003263621 and parameters: {'category': 'foo', 'x1': 9.684367249984621}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,275] Trial 20 finished with value: 1.9472463883739826 and parameters: {'category': 'foo', 'x1': 0.6045622950579332}. Best is trial 11 with value: 63.81133568419859.
[I 2022-06-16 11:43:25,277] Trial 21 finished with value: 63.86901591705744 and parameters: {'category': 'foo', 'x1': 9.991809301845073}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,280] Trial 22 finished with value: 47.970539401366366 and parameters: {'category': 'foo', 'x1': 8.926076768370848}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,282] Trial 23 finished with value: 25.446496546728522 and parameters: {'category': 'foo', 'x1': 7.0444520561433155}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,284] Trial 24 finished with value: 63.65625553605348 and parameters: {'category': 'foo', 'x1': 9.978487045552777}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,287] Trial 25 finished with value: 27.202169723029566 and parameters: {'category': 'foo', 'x1': 7.215569932713928}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,289] Trial 26 finished with value: -0.31831227354973746 and parameters: {'category': 'bar', 'x2': -5.564191699291772}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,291] Trial 27 finished with value: 48.49249048129746 and parameters: {'category': 'foo', 'x1': 8.963654965698506}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,294] Trial 28 finished with value: 48.22747947576172 and parameters: {'category': 'foo', 'x1': 8.944600742718167}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,297] Trial 29 finished with value: -24.584445666830728 and parameters: {'category': 'bar', 'x2': -0.041729569009902434}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,299] Trial 30 finished with value: 0.19249940833440185 and parameters: {'category': 'foo', 'x1': 1.5612524548964384}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,302] Trial 31 finished with value: 63.736800148253934 and parameters: {'category': 'foo', 'x1': 9.983533061762438}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,304] Trial 32 finished with value: 62.43550399886475 and parameters: {'category': 'foo', 'x1': 9.901614012267668}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,307] Trial 33 finished with value: 49.20440007171121 and parameters: {'category': 'foo', 'x1': 9.01458481107123}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,310] Trial 34 finished with value: 30.598330349863218 and parameters: {'category': 'foo', 'x1': 7.531575756496807}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,312] Trial 35 finished with value: 52.31168547494971 and parameters: {'category': 'foo', 'x1': 9.232681762316776}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,315] Trial 36 finished with value: -7.3249841467316985 and parameters: {'category': 'bar', 'x2': -7.706470791775093}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,317] Trial 37 finished with value: 63.45722838945394 and parameters: {'category': 'foo', 'x1': 9.966004543650094}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,320] Trial 38 finished with value: -2.2662175132514135 and parameters: {'category': 'bar', 'x2': -3.494603868328534}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,322] Trial 39 finished with value: 1.0788073758518255 and parameters: {'category': 'foo', 'x1': 3.038656524483347}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,325] Trial 40 finished with value: 41.965785770735714 and parameters: {'category': 'foo', 'x1': 8.478100475504815}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,327] Trial 41 finished with value: 54.91065619744565 and parameters: {'category': 'foo', 'x1': 9.410172480951145}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,330] Trial 42 finished with value: 62.360322919664895 and parameters: {'category': 'foo', 'x1': 9.896855255078751}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,333] Trial 43 finished with value: 46.45605217871198 and parameters: {'category': 'foo', 'x1': 8.81586767614454}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,335] Trial 44 finished with value: 63.506839532449206 and parameters: {'category': 'foo', 'x1': 9.969117864133345}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,338] Trial 45 finished with value: 34.351727969581155 and parameters: {'category': 'foo', 'x1': 7.861034718339515}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,341] Trial 46 finished with value: 40.662885003741636 and parameters: {'category': 'foo', 'x1': 8.376745643644698}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,343] Trial 47 finished with value: 10.495243549606926 and parameters: {'category': 'foo', 'x1': 5.239636329838108}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,346] Trial 48 finished with value: 53.76780332344284 and parameters: {'category': 'foo', 'x1': 9.33265322536412}. Best is trial 21 with value: 63.86901591705744.
[I 2022-06-16 11:43:25,348] Trial 49 finished with value: -3.405327722423339 and parameters: {'category': 'bar', 'x2': -6.845353007536319}. Best is trial 21 with value: 63.86901591705744.
/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/bug_matplotlib_plot_contour.py:28: ExperimentalWarning:

plot_contour is experimental (supported from v2.2.0). The interface can change in the future.

[W 2022-06-16 11:43:27,726] Output figures of this Matplotlib-based `plot_contour` function would be different from those of the Plotly-based `plot_contour`.
Traceback (most recent call last):
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/bug_matplotlib_plot_contour.py", line 33, in <module>
    main()
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/bug_matplotlib_plot_contour.py", line 28, in main
    matplotlib_visualization.plot_contour(study)
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/venv/lib/python3.9/site-packages/optuna/_experimental.py", line 72, in wrapper
    return func(*args, **kwargs)
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/venv/lib/python3.9/site-packages/optuna/visualization/matplotlib/_contour.py", line 102, in plot_contour
    return _get_contour_plot(study, params, target, target_name)
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/venv/lib/python3.9/site-packages/optuna/visualization/matplotlib/_contour.py", line 170, in _get_contour_plot
    cs = _generate_contour_subplot(
  File "/Users/c-bata/go/src/github.com/c-bata/optuna-visualization-regression-tests/venv/lib/python3.9/site-packages/optuna/visualization/matplotlib/_contour.py", line 412, in _generate_contour_subplot
    ax.set_xlim(x_values_range[0], x_values_range[1])
IndexError: list index out of range
```

### Steps to reproduce

1. Execute following Python script

```python
import optuna
import optuna.visualization as plotly_visualization
import optuna.visualization.matplotlib as matplotlib_visualization
import matplotlib.pylab as plt


def objective_single_dynamic(trial: optuna.Trial) -> float:
    category = trial.suggest_categorical("category", ["foo", "bar"])
    if category == "foo":
        return (trial.suggest_float("x1", 0, 10) - 2) ** 2
    else:
        return -((trial.suggest_float("x2", -10, 0) + 5) ** 2)


def main():
    study = optuna.create_study(
        study_name="A single-objective study with 3-dimensional dynamic search space",
        direction="maximize",
    )

    study.optimize(objective_single_dynamic, n_trials=50)

    # Plotly
    plotly_fig = plotly_visualization.plot_contour(study)
    plotly_fig.write_image("plotly.png")

    # Matplotlib
    matplotlib_visualization.plot_contour(study). # an exception is raised here!
    plt.savefig("matplotlib.png", bbox_inches="tight")


if __name__ == '__main__':
    main()
```

### Additional context (optional)

I detected this bug by https://github.com/c-bata/optuna-visualization-regression-tests.

Here is an output.
https://c-bata.link/optuna-visualization-regression-tests/plot_contour.html

<img width="1346" alt="Screenshot 2022-06-16 11 50 15" src="https://user-images.githubusercontent.com/5564044/173980609-8f2de5f1-f3df-4186-8bb4-bdc9ccbb60a7.png">

**Comment by HideakiImamura** [comment]
This will be fixed in #3682.

**Comment by c-bata** [comment]
I confirmed that this bug is fixed at #3682.
https://optuna.github.io/optuna-visualization-regression-tests/plot_contour.html
